# Rhea_chip 
A Simple pipeline for the analysis of human data generated by NGS approaches.



------
Rhea_chip is a simple pipeline for the analysis of human data generated by NGS approaches. With Rhea_chip you can now have full suite of NGS tools up and running on any high end workstation in an afternoon. 

**This pipeline will provide a workflow as follows:**

From the fastq files generated from the Illumina pipeline, all of them were first processed by filtering adaptor sequences and removing low-quality reads. Cleaned reads were aligned to the GRCh37/hg19 human reference genome (UCSC Genome Browser) with aln and sampe tools of the Burrows-Wheeler Aligner (BWA)(Li & Durbin, 2010). The resulting bam files were sorted by using the SortSam tools which is part of the Picard toolkit  (http://sourceforge.net/projects/picard/). The Picard MergeSamFiles tool was performed to merge the sorted bam files which belong to one sample into one bam file. Potential PCR artifacts were marked with the Picard MarkDuplicates tool. Notably, alignment to the human reference particularly in areas around small base pair insertions and deletions (indels) is often imperfect which may misalign. To facilitate the identification of Indels, alignments were refined with the Genome Analysis Toolkit(GATK) RealignerTargetCreator and IndelRealigner tools. When used together, these tools use the full alignment context to determine whether an Indel exists. Quality scores for sequencing reads were recalibrated using BaseRecalibrator and PrintReads from GATK to more closely match the actual probability of mismatching the reference and to correct any variation in quality between machine cycle and sequence context(DePristo et al., 2011). An index file for each BAM file was generated using the SAMTools index function. Variants were called using the HaplotypeCaller from GATK (https://www.broadinstitute.org/gatk/). HaplotypeCaller calls both single nucleotide polymorphisms and indels using de novo assembly of haplotypes in the target region. After variants were called and exported in variant call format (VCF), we performed the GATK tool VariantFiltration to filter variants. SNV and indel VCF files were combined using GATK CombineVariants tool.

------

### Environment

Of course, pipeline software is only useful if it can be executed. You should find them in [here][1]

### INPUT

Short sequencing reads in fastq format (Illumina platform). Note: Other platforms, notably Pacific Biosciences-generated long reads, should be acceptable input; however, since data are pre-assembled on-instrument it is likely that the Assemble step (below) is unnecessary.
According to our pipeline requirements, you need to prepare a sample.list, and a configuration file (config.json)

### USAGE

```bash
Rhea_Chip.py [-h] [--version] --samplelist [samplelist] --config [config.json] [options]
```

```html
Usage: 

A Simple pipeline for the analysis of human data generated by NGS approaches.

Options:
  --version             show program's version number and exit
  -h, --help            show this help message and exit

  Expected arguments:
    Caution: These parameters are necessary for Rhea_Chip.

    -s FILE, --samplelist=FILE
                        The sample list file
                        Column 1:    sample name
                        Column 2:    family name
                        Column 3:    library
                        Column 4:    FQ file [formact:lane1/fq1:lane1/adapter1;lane1/fq2:lane1/adapter2|lane2/fq1:lane2/adapter1;lane2/fq2:lane2/adapter2|...]
                        Column 5:    case/control
    -c FILE, --config=FILE
                        The config file, contain all specific settings in a
                        json file

  Optional arguments:
    Caution: If you do not set these parameters in addition, Rhea_Chip
    will select the default value.

    -o OUTDIR, --outdir=OUTDIR
                        The output dir [ default:
                        /Users/huang/Desktop/Rhea_Chip ]

Author: joeyhwong@hotmail.com
```

### FLOW CHART

```flow
st=>start: Prepare sample.list and config.json
op=>operation: Run Rhea_chip
cond=>condition: Operating normally?
e=>end: In order to gradually run each script

st->op->cond
cond(yes)->e
cond(no)->op
```

### Structure

```html
|-- Rhea_Chip.py  <!-- main program -->
|-- bin <!-- Some required script -->
|   |-- bam_to_chip_gender.py <!-- Gender Prediction -->
|   |-- depthQC.py <!-- Quality Control base on deoth and coverage -->
|   `-- vcf_phasing.py <!-- VCF phasing -->
|-- config <!-- Configuration file -->
|   `-- config.json
|-- dataset <!-- Some required dataset -->
|   |-- dbsnp_138.hg19.vcf.gz <!-- dbsnp (ftp://gsapubftp-anonymous@ftp.broadinstitute.org/bundle/2.8/hg19/dbsnp_138.hg19.vcf.gz) -->
|   |-- dbsnp_138.hg19.vcf.gz.tbi
|   |-- hg19.dbref.clean.gz <!-- ncbi dbref -->
|   `-- hg19.dbref.clean.gz.tbi
|--------- The following is the test datas ---------  
|-- sample.list <!-- Sample message file [Test] -->
|-- SampleTest <!-- Test... -->
|   |-- alignment <!-- Alignment results -->
|   |   `-- 2015003 <!-- samplename -->
|   |       |-- 2015003.final.bam <!-- Alignment file -->
|   |       `-- 2015003.final.bam.bai
|   |-- doc <!-- Report Documents -->
|   |   |-- 2015.12.07.Flank.phig.norm.bed <!-- Flank BedFile corrected -->
|   |   |-- 2015.12.07.Target.phig.norm.bed <!-- Target BedFile corrected -->
|   |   |-- QC <!-- QC -->
|   |   |   `-- 2015003 <!-- Family Name -->
|   |   |       `-- 2015003 <!-- Sample Name -->
|   |   |           |-- 2015003.lane_1.gender
|   |   |           |-- 2015003.lane_2.gender
|   |   |           |-- Base_distributions_by_read_position_1.txt
|   |   |           |-- Base_distributions_by_read_position_2.txt
|   |   |           |-- Base_quality_value_distribution_by_read_position_1.txt
|   |   |           |-- Base_quality_value_distribution_by_read_position_2.txt
|   |   |           |-- Basic_Statistics_of_Sequencing_Quality.txt
|   |   |           |-- Distribution_of_Q20_Q30_bases_by_read_position_1.txt
|   |   |           |-- Distribution_of_Q20_Q30_bases_by_read_position_2.txt
|   |   |           |-- Statistics_of_Filtered_Reads.txt
|   |   |           |-- bed_anno
|   |   |           |   |-- 2015003.Flank.bed.anno
|   |   |           |   `-- 2015003.Target.bed.anno
|   |   |           |-- depth_anno
|   |   |           |   |-- Flank
|   |   |           |   |   |-- 2015003.RawDepth.chrom.stat
|   |   |           |   |   |-- 2015003.RawDepth.depth
|   |   |           |   |   |-- 2015003.RawDepth.gz
|   |   |           |   |   |-- 2015003.RawDepth.gz.tbi
|   |   |           |   |   |-- 2015003.RawDepth.stat
|   |   |           |   |   |-- 2015003.RmdupDepth.chrom.stat
|   |   |           |   |   |-- 2015003.RmdupDepth.depth
|   |   |           |   |   |-- 2015003.RmdupDepth.gz
|   |   |           |   |   |-- 2015003.RmdupDepth.gz.tbi
|   |   |           |   |   |-- 2015003.RmdupDepth.stat
|   |   |           |   |   `-- 2015003.RmdupDepth.uncover.anno
|   |   |           |   `-- Target
|   |   |           |       |-- 2015003.RawDepth.chrom.stat
|   |   |           |       |-- 2015003.RawDepth.depth
|   |   |           |       |-- 2015003.RawDepth.gz
|   |   |           |       |-- 2015003.RawDepth.gz.tbi
|   |   |           |       |-- 2015003.RawDepth.stat
|   |   |           |       |-- 2015003.RmdupDepth.chrom.stat
|   |   |           |       |-- 2015003.RmdupDepth.depth
|   |   |           |       |-- 2015003.RmdupDepth.gz
|   |   |           |       |-- 2015003.RmdupDepth.gz.tbi
|   |   |           |       |-- 2015003.RmdupDepth.stat
|   |   |           |       `-- 2015003.RmdupDepth.uncover.anno
|   |   |           `-- graph_exon <!-- Graphs of Exon depth and coverage -->
|   |   |-- Rhea_chip.config <!-- Configuration file -->
|   |   |-- Rhea_chip.log <!-- log -->
|   |   `-- sample.list2015.12.07 <!-- Sample message file -->
|   |-- fq <!-- fastqs -->
|   |   |-- clean_data <!-- Fqs after clean adapters -->
|   |   |   `-- 2015003
|   |   |       |-- 2015003_lane_1
|   |   |       |   |-- 2015003.lane_1.1.fq.gz
|   |   |       |   `-- 2015003.lane_1.2.fq.gz
|   |   |       `-- 2015003_lane_2
|   |   |           |-- 2015003.lane_2.1.fq.gz
|   |   |           `-- 2015003.lane_2.2.fq.gz
|   |   `-- raw_data <!-- Raw Fqs, Just a soft link -->
|   |       `-- 2015003
|   |           |-- 2015003.lane_1.fastq_1.adapter.list.gz -> /share/fqdata82/HiSeq2500_rapid/F14ZQSQSSJ1554_HUMjboX/JSZLSPEP00000682/150807_I176_FCHYCHKADXX_L1_HUMhgfXQDDAAAPEI-9/1.adapter.list.gz
|   |           |-- 2015003.lane_1.fastq_1.fq.gz -> /share/fqdata82/HiSeq2500_rapid/F14ZQSQSSJ1554_HUMjboX/JSZLSPEP00000682/150807_I176_FCHYCHKADXX_L1_HUMhgfXQDDAAAPEI-9/150807_I176_FCHYCHKADXX_L1_HUMhgfXQDDAAAPEI-9_1.fq.gz
|   |           |-- 2015003.lane_1.fastq_2.adapter.list.gz -> /share/fqdata82/HiSeq2500_rapid/F14ZQSQSSJ1554_HUMjboX/JSZLSPEP00000682/150807_I176_FCHYCHKADXX_L1_HUMhgfXQDDAAAPEI-9/2.adapter.list.gz
|   |           |-- 2015003.lane_1.fastq_2.fq.gz -> /share/fqdata82/HiSeq2500_rapid/F14ZQSQSSJ1554_HUMjboX/JSZLSPEP00000682/150807_I176_FCHYCHKADXX_L1_HUMhgfXQDDAAAPEI-9/150807_I176_FCHYCHKADXX_L1_HUMhgfXQDDAAAPEI-9_2.fq.gz
|   |           |-- 2015003.lane_2.fastq_1.adapter.list.gz -> /share/fqdata82/HiSeq2500_rapid/F14ZQSQSSJ1554_HUMjboX/JSZLSPEP00000682/150807_I176_FCHYCHKADXX_L2_HUMhgfXQDDAAAPEI-9/1.adapter.list.gz
|   |           |-- 2015003.lane_2.fastq_1.fq.gz -> /share/fqdata82/HiSeq2500_rapid/F14ZQSQSSJ1554_HUMjboX/JSZLSPEP00000682/150807_I176_FCHYCHKADXX_L2_HUMhgfXQDDAAAPEI-9/150807_I176_FCHYCHKADXX_L2_HUMhgfXQDDAAAPEI-9_1.fq.gz
|   |           |-- 2015003.lane_2.fastq_2.adapter.list.gz -> /share/fqdata82/HiSeq2500_rapid/F14ZQSQSSJ1554_HUMjboX/JSZLSPEP00000682/150807_I176_FCHYCHKADXX_L2_HUMhgfXQDDAAAPEI-9/2.adapter.list.gz
|   |           `-- 2015003.lane_2.fastq_2.fq.gz -> /share/fqdata82/HiSeq2500_rapid/F14ZQSQSSJ1554_HUMjboX/JSZLSPEP00000682/150807_I176_FCHYCHKADXX_L2_HUMhgfXQDDAAAPEI-9/150807_I176_FCHYCHKADXX_L2_HUMhgfXQDDAAAPEI-9_2.fq.gz
|   |-- javatmp <!-- temp files when running the java program -->
|   |-- script <!-- Pipeline -->
|   |   `-- 2015003
|   |       |-- Step1.2015003.lane_1.filter.sh
|   |       |-- Step1.2015003.lane_2.filter.sh
|   |       |-- Step2.2015003.lane_1.align.sh
|   |       |-- Step2.2015003.lane_2.align.sh
|   |       |-- Step3.2015003.merge.sh
|   |       |-- Step4.2015003.dupmark.sh
|   |       |-- Step5.2015003.RealignerTargetCreator.sh
|   |       |-- Step6.1.2015003.vcf_calling.sh
|   |       |-- Step6.2.2015003.depthQC.sh
|   `-- variation <!-- Results -->
|       `-- 2015003
|           |-- 2015003-2.final.vcf.gz <!-- Final VCF -->
|           |-- 2015003-2.final.vcf.gz.tbi
|           |-- 2015003.final.vcf.bak
|           |-- 2015003.final.vcf.idx
|           |-- 2015003.indel.vcf
|           |-- 2015003.indel.vcf.idx
|           |-- 2015003.snv.vcf
|           |-- 2015003.snv.vcf.idx
|           |-- 2015003.vcf
|           |-- 2015003.vcf.idx
`-- README.md
```


  [1]: ./config/config.json
