#Objective
Install all software packages required to follow the GATK Best Practices.

#Prerequisites
To follow these instructions, you will need to have a basic understanding of the meaning of the following words and command-line operations. If you are unfamiliar with any of the following, you should consult a more experienced colleague or your systems administrator if you have one. There are also many good online tutorials you can use to learn the necessary notions.

> * Basic Unix environment commands
> * Binary / Executable
> * Compiling a binary
> * Adding a binary to your path
> * Command-line shell, terminal or console

#Software library
You will also need to have access to an ANSI compliant C++ compiler and the tools needed for normal compilations (make, shell, the standard library, tar, gunzip). These tools are usually pre-installed on Linux/Unix systems. **On MacOS X, you may need to install the MacOS Xcode tools. See [https://developer.apple.com/xcode/](https://developer.apple.com/xcode/) for relevant information and software downloads. **The XCode tools are free but an AppleID may be required to download them.

Rhea_Chip requires python (version >= 2.7) and Java (version >= 1.7). All Linux/Unix and MacOS X systems should have a JRE pre-installed, but the version may vary. To test your python / Java version, run the following command in the shell:

```
python -V
java -version
```


#Software packages
> * BWA
> * SAMtools
> * Picard
> * Genome Analysis Toolkit (GATK)
> * SOAPnumke
> * snpEff
> * python libraries

*Note that the version numbers of packages you download may be different than shown in the source.list. If so, software will re-install while you install RheaChip.*

Installation

```
export Rhea_Chip_Home=$HOME # $Home or anyware
python setup install
python Rhea_Chip/Rhea_Chip/install.py
```


## Important note

We do not provide all database files which will be used in our analysis, everyone can build the use databases refer on our demo.
RheaChip is Open-source, but some softwares in our pipeline may be commercial use, THIS IS WHAT WE NEED TO PAY ATTENTION !!

### Configuration file

Of course, pipeline software is only useful if it can be executed. You can find them in [here][1]

### INPUT

Short sequencing reads in fastq format (Illumina platform). Note: Other platforms, notably Pacific Biosciences-generated long reads, should be acceptable input; however, since data are pre-assembled on-instrument it is likely that the Assemble step (below) is unnecessary.
According to our pipeline requirements, you need to prepare a sample.list, and a configuration file.

### USAGE

```
=> Rhea_Chip_pipeline.py
Usage: Rhea_Chip_pipeline.py [-h] [--version] --samplelist [sample list] --config [config file] [options]

A Simple pipeline for the analysis of human data generated by NGS approaches.

Options:
  --version             show program's version number and exit
  -h, --help            show this help message and exit

  Expected arguments:
    Caution: These parameters are necessary for Rhea_Chip.

    -s FILE, --samplelist=FILE
                        The sample list file
                        Column 1:    sample name
                        Column 2:    family name
                        Column 3:    library
                        Column 4:    FQ file [formact:
                        lane1/fq1:lane1/adapter1;lane1/fq2:lane1/adapter2|
                        lane2/fq1:lane2/adapter1;lane2/fq2:lane2/adapter2|...]
                        Column 5:    case/control
    -c FILE, --config=FILE
                        The config file, contain all specific settings

  Optional arguments:
    Caution: If you do not set these parameters in addition, Rhea_Chip
    will select the default value.

    -o OUTDIR, --outdir=OUTDIR
                        The output dir [ default: `pwd` ]
    -p PROJECT_NAME, --project_name=PROJECT_NAME
                        set your project name

Joey Hwong <joeyhwong@hotmail.com>

```

### FLOW CHART

```flow
st=>start: Prepare sample.list and config.conf
op=>operation: Run Rhea_chip
cond=>condition: Operating normally?
e=>end: In order to gradually run each script

st->op->cond
cond(yes)->e
cond(no)->op
```

### Structure
```
alignment: bam files
annotation: final result
cnv: cnv scripts and cnv result
doc: QC
fq: fastq
javatmp: java tmp
script: analysis scripts
variation : vcf
```



[1]: ./config/config/Rhea_Chip.conf
